### ISSUE TITLE
IDE0060 - Remove unused parameter for internal method referenced by a delegate in another type

### ISSUE THREAD

**Issue by bartdesmet** [dotnet/roslyn#32851]
This warning fires incorrectly when the method is used in a method group conversion from another type, as shown below with `Bar.Foo` referenced from `Baz`.

```csharp
class Baz
{
    private static readonly Action<int> s_foo = Bar.Foo;
}

class Bar
{
    // IDE0060 - Remove unused parameter
    internal static void Foo(int unused) { }
}

class Qux
{
    private static readonly Action<int> s_foo = Foo;

    // No warning
    internal static void Foo(int unused) { }
}
```

**Comment by mavasani** [comment]
This is almost identical to https://github.com/dotnet/roslyn/issues/32228#issuecomment-453700464: 

_We do have logic in the analyzer to track and bail out in presence of method references (as opposed to invocations), but the tracking is local to a type (and its nested types), as that is the most common case and addressing the case in Test2 above would require us to convert the analyzer into a compilation end analyzer to track address captures within compilation, and it would still not work if the address capture happens in a different project.
So, unfortunately this is by design based on our current stance to avoid compilation end analyzers in the IDE, you would need to add suppressions. If this leads to lot of noise or gets more customer requests, then we can revisit how to avoid this false positive._

Marking this as tentatively by design, and also needs design review so we can confirm this behavior in a design meeting.

**Comment by mavasani** [comment]
Tagging @sharwell - can we bring this up in the design meeting today?

**Comment by bartdesmet** [comment]
Some context on where this came up.

We have a compiler for efficient type-directed data structure deserialization. Some class builds a parser by emitting lexer code and by emitting calls to parse actions. The latter have a delegate contract to accept a cursor, a length, and a context object that's threaded through. Various files across the code base provide such actions for different types, some of which don't care about e.g. the context object. The warning fires for these methods, but they have to match the delegate signature. A central catalog file registers all of these methods in some dictionary whose value type is the delegate type. It are those use sites that are not observed by the analyzer.

I've worked around it by suppressing the warning for now.

**Comment by bartdesmet** [comment]
FWIW, I noodled around with some other examples using events, which are merely special cases for referencing methods through delegates. I was surprised to find the special case handling for traditional event signatures (even though `EventHandler<T>` no longer has a constraint for `T` to derive from `EventArgs`).

```csharp
class Baz
{
    public event EventHandler<EventArgs> E;
    public event EventHandler<string> F;
    public event Action<string> G;

    public void Foo()
    {
        E += Bar.Foo1;
        F += Bar.Foo2;
        G += Bar.Foo3;
    }
}

class Bar
{
    internal static void Foo1(object sender, EventArgs e) { } // No warning?
    internal static void Foo2(object sender, string e) { } // IDE0060 - Remove unused parameter
    internal static void Foo3(string s) { } // IDE0060 - Remove unused parameter
}
```

**Comment by CyrusNajmabadi** [comment]
Aside: this seems unfortunate from the language level.  it would be nice if you could remove those parameters and still attach them to delegates/events.  We already allow and understand co/contravariance when assigning this stuff.  Allowing fewer params just fits into that.  

Also: `// No warning?`.  I think it *may* be the case that 'EventArgs' is just special cased.

**Comment by mavasani** [comment]
>  I think it may be the case that 'EventArgs' is just special cased.

Yes, that is exactly the case. The analyzer special cases methods with event handler signature and also bails out on parameters if the containing method is used as a delegate within the type. These two are most common cases for using methods as delegates, and hence handled in the analyzer. The other cases you mentioned are much rare, and handling them in the analyzer would need to expand analyzer's analysis scope to entire project (no IVT case) or entire solution, which make them unsuitable for a live analyzer.

**Comment by bartdesmet** [comment]
> this seems unfortunate from the language level

Agreed. Anonymous methods with `delegate { … }` syntax already support omitting all parameters if unused (but lambdas don't). Discard `_` has been a nice addition too, to convey one's intent of ignoring something (unlike having to resort to `#pragma` suppressions which are very invasive and verbose), but `_` is already a valid identifier for a parameter. And dare I mention C++ where you can omit parameter names altogether if you're not gonna use them.

> Allowing fewer params

Seems like this would only work for trailing parameters, while I may have an unused parameter in the middle. Some options that feel clean to me would be `void Foo(int a, bool _, char c)` or `void Foo(int a, bool, char c)`, but the former suffers from `_` being a valid identifier today, and the latter is brand new syntax.

**Comment by bartdesmet** [comment]
> expand analyzer's analysis scope to entire project

The challenge is understood, and suppression is feasible here. This said, suppression is incredibly verbose, so I often resort to `GlobalSuppressions.cs` to get it out of view. As Cyrus pointed out, maybe worth discussion at the LDM level to see if there's value in exploring ways to indicate intent of omitting parameters, akin to `_`?

**Comment by mavasani** [comment]
@bartdesmet I like your and @CyrusNajmabadi's proposal. Can you please file a separate compiler/language feature request with context from this issue for LDM? Thanks!

**Comment by CyrusNajmabadi** [comment]
@mavasani we might also want a naming convention for things we just ignore. i.e. `int _` or `int _1` `int _2` `int _3` etc.  That way users can have a reasonable way to say "yes... i know this parameter is here.  but please ignore it", without the huge `suppression` directive you need otherwise.

**Comment by YairHalberstadt** [comment]
It would probably be quite simple to suppress the warning automatically when the unused parameter is named `_`, and this should cover a reasonable percentage of cases.

**Comment by mavasani** [comment]
@CyrusNajmabadi @YairHalberstadt - that is a very good idea, that is also a good solution for #32228. Let me create a PR. Thanks!

**Comment by mavasani** [comment]
@YairHalberstadt - seems like we already handle `_`, but I agree with Cyrus that we want to allow multiple unused parameters, so bailing out on `_1`, `_2`, etc. is the right approach.
